datalad/interface/utils.py

```diff
@@ -667,3 +667,84 @@ def _discover_trace_to_known(path, trace, spec):
         if not isdir(p):
             continue
         _discover_trace_to_known(p, trace, spec)
+
+
+def filter_unmodified(content_by_ds, refds, since):
+    """Filter per-dataset path specifications based on modification history.
+
+    This function takes a path specification dictionary, as produced by
+    `Interface._prep()` and filters it such that only that subset of paths
+    remains in the dictionary that correspdong to the set of changes in
+    the given reference dataset since a given state.
+
+    The change set is traced across all related subdatasets, i.e. if a submodule
+    in the reference dataset is reported as modified then all paths for any given
+    subdataset in the modified one are tested for changes too (based on the
+    state at which the parent dataset reports a change in the subdataset), and so
+    on.
+
+    In doing so, not only unmodified given paths are removed, but also modified
+    given paths are replaced by the set of actually modified paths within them.
+
+    Only commited changes are considered!
+
+    Parameters
+    ----------
+    content_by_ds : dict
+      Per-dataset path specifications, as produced ,for example, by
+      `Interface._prep()`
+    refds : Dataset
+      Instance of a reference dataset for which to determine the initial
+      change set
+    since : state
+      Any commit-ish/tree-ish supported by Git (tag, commit, branch, ...).
+      Changes between this given state and the most recent commit are
+      evaluated.
+
+    Returns
+    -------
+    dict
+      Filtered path spec dictionary. The output is guaranteed to only contain
+      paths to modified, and presently existing components of subdatasets
+      of the given reference dataset (and itself).
+    """
+    # life is simple: we diff the base dataset, and kill anything that
+    # does not start with something that is in the diff
+    # we cannot really limit the diff paths easily because we might get
+    # or miss content (e.g. subdatasets) if we don't figure out which ones
+    # are known -- and we don't want that
+    diff = refds.repo.repo.commit().diff(since)
+    # get all modified paths (with original? commit) that are still
+    # present
+    modified = dict((opj(refds.path, d.b_path),
+                    # talk to @yaricoptic about the next line
```

Comment:
I got it now!  but meanwhile here is probably a handy demo case:
```shell
$> d=/tmp/d; rm -rf $d; mkdir $d; cd $d; git init; echo 1 > modified; echo 2 > removed; echo 1 > renamed; git add modified removed renamed; git commit -m 'initial commit'; echo 11 >> modified; rm removed; echo 33 > added; git mv renamed renamed2; git add added; git commit -m "removed added etc" -a  ; 
Initialized empty Git repository in /tmp/d/.git/          
[master (root-commit) 7a2f2c1] initial commit
 3 files changed, 3 insertions(+)
 create mode 100644 modified
 create mode 100644 removed
 create mode 100644 renamed
[master 28fd456] removed added etc
 4 files changed, 2 insertions(+), 1 deletion(-)
 create mode 100644 added
 delete mode 100644 removed
 rename renamed => renamed2 (100%)

$> datalad install -d . ///labs/haxby/raiders; cd raiders; git reset --hard HEAD^; cd ..; git commit -m 'downgrade raiders' -a;
[INFO   ] Installing subdataset from 'http://datasets.datalad.org/labs/haxby/raiders' at: http://datasets.datalad.org/labs/haxby/raiders 
[INFO   ] Saving possible changes to <Dataset path=/tmp/d> - installed subdataset: raiders 
                                                                                                                          
[INFO   ] Saving possible changes to <Dataset path=/tmp/d> - installed subdataset: raiders 
1 installed item is available at
<Dataset path=/tmp/d/raiders>
```

so let's check:
```
$> python -c 'from git import Repo; r = Repo("."); print [((d.a_path, bool(d.a_blob)), (d.b_path, bool(d.b_blob))) for d in r.commit().diff("HEAD^^")]'
[((u'.gitmodules', True), (u'.gitmodules', False)), ((u'raiders', True), (u'raiders', False))]
```
so it seems that b's blob is null... and altogether:
```shell
$> python -c 'from git import Repo; r = Repo("."); print [((d.a_path, bool(d.a_blob)), (d.b_path, bool(d.b_blob))) for d in r.commit().diff("HEAD^^^")]' 
[((u'.gitmodules', True), (u'.gitmodules', False)), ((u'added', True), (u'added', False)), ((u'modified', True), (u'modified', True)), ((u'raiders', True), (u'raiders', False)), ((u'removed', False), (u'removed', True)), ((u'renamed2', True), (u'renamed', True))]
```

So if the question was -- `why b` -- then what I have missed in the (absent) documentation of `_get_changed_datasets` is that it returns a dictionary for paths with values being previous states, so I could pass those as "since" into the corresponding sub-dataset `publish` invocations, and here you also pass recursively into `filter_unmodified`. That was "commented" on as 
```# if we were returned a dict, we got subds_prev_hexsha```
in `publish` logic.  Also in that function I was explicitly checking for no renames (so paths were the same on a and b) for now, so you might improve upon that here while at it ;)